inference-engine/src/auto_plugin/auto_plugin.cpp

```diff
@@ -0,0 +1,311 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+#include <string>
+#include <vector>
+#include <memory>
+#include <map>
+#include <unordered_map>
+#include <unordered_set>
+
+#include <ie_metric_helpers.hpp>
+#include <auto_plugin/auto_config.hpp>
+#include <ie_core.hpp>
+#include <threading/ie_executor_manager.hpp>
+#include <ie_algorithm.hpp>
+
+#include "auto_plugin.hpp"
+
+// ------------------------------AutoInferencePlugin----------------------------
+namespace AutoPlugin {
+    using namespace InferenceEngine;
+namespace {
+    std::map<std::string, std::string> mergeConfigs(std::map<std::string, std::string> config,
+                                                    const std::map<std::string, std::string> & local) {
+        for (auto && kvp : local) {
+            config[kvp.first] = kvp.second;
+        }
+        return config;
+    }
+}  // namespace
+
+AutoInferencePlugin::AutoInferencePlugin() {
+    _pluginName = "AUTO";
+    RegisterPolicy(SelectDevicePolicy::STATIC);
+}
+
+ExecutableNetworkInternal::Ptr AutoInferencePlugin::LoadExeNetworkImpl(const CNNNetwork &network,
+                                                                       const std::map<std::string, std::string>& config) {
+    if (GetCore() == nullptr) {
+        IE_THROW() << "Please, work with AUTO device via InferencEngine::Core object";
+    }
+
+    if (network.getFunction() == nullptr) {
+        IE_THROW() << "AUTO device supports just ngraph network representation";
+    }
+
+    auto fullConfig = mergeConfigs(_config, config);
+    auto deviceChoiceConfig = fullConfig.find(AutoConfigParams::KEY_AUTO_DEVICE_CHOICE);
+    if (deviceChoiceConfig == fullConfig.end()) {
+        auto deviceChoice = GetDeviceChoice();
+        fullConfig.emplace(AutoConfigParams::KEY_AUTO_DEVICE_CHOICE, deviceChoice);
+    }
+    auto metaDevices = ParseMetaDevices(fullConfig[AutoConfigParams::KEY_AUTO_DEVICE_CHOICE], fullConfig);
+
+    // collect the settings that are applicable to the devices we are loading the network to
+    std::unordered_map<std::string, InferenceEngine::Parameter> autoNetworkConfig;
+    autoNetworkConfig.insert(*fullConfig.find(AutoConfigParams::KEY_AUTO_DEVICE_CHOICE));
+
+    auto scheduleType = ParseScheduleType(fullConfig[AutoConfigParams::KEY_AUTO_SCHEDULE_TYPE]);
+    auto optCap = GetOptimizationCapabilities();
+
+    ExecutableNetwork executableNetwork;
+    DeviceInformation selectedDevice {};
+    while (!metaDevices.empty()) {
+        selectedDevice = _policies[scheduleType]->SelectDevice(network, metaDevices, optCap);
+        try {
+            auto deviceQr = GetCore()->QueryNetwork(network, selectedDevice.deviceName, selectedDevice.config);
```

Comment:
Hi Maxim, this logic orignally is for the handling of loading network failure (if dGPU failed, select iGPU, if iGPU failed, select CPU, we need load network every time). we thought QueryNetwork could quickly check if the network could be loaded to that device, since the loading may cost lots of time (such as memory copy and network pipeline creation). But of course, we don't know how much benefit this logic could achieve.
@wangleis what do you think?

Reply:
I will delete it for this first version, may discuss in later PR.